libsolutil/JSON.cpp

```diff
@@ -22,135 +22,218 @@
-/// \param _input JSON input string
-/// \param _json [out] resulting JSON object
-/// \param _errs [out] Formatted error messages
-/// \return \c true if the document was successfully parsed, \c false if an error occurred.
-bool parse(Json::CharReaderBuilder& _builder, std::string const& _input, Json::Value& _json, std::string* _errs)
+std::string formatLikeJsoncpp(std::string const& _dumped, JsonFormat const& _format)
 {
-	std::unique_ptr<Json::CharReader> reader(_builder.newCharReader());
-	return reader->parse(_input.c_str(), _input.c_str() + _input.length(), &_json, _errs);
+	uint32_t indentLevel = 0;
+	std::stringstream reformatted;
+	bool inQuotes = false;
+	for (size_t i = 0; i < _dumped.size(); ++i)
+	{
+		char c = _dumped[i];
+		bool emptyThing = false;
+
+		if (c == '"' && (i == 0 || _dumped[i - 1] != '\\'))
+			inQuotes = !inQuotes;
+
+		if (!inQuotes)
+		{
+			if (i < _dumped.size() - 1)
+			{
+				char nc = _dumped[i + 1];
+				if ((c == '[' && nc == ']') || (c == '{' && nc == '}'))
+					emptyThing = true;
+			}
+			if (c == '[' || c == '{')
+			{
+				if (i > 0 && _dumped[i - 1] != '\n')
+					if (!emptyThing)
+						reformatted << '\n' << std::string(indentLevel * _format.indent, ' ');
+				indentLevel++;
+			}
+			else if (c == ']' || c == '}')
+			{
+				indentLevel--;
+				if (i + 1 < _dumped.size() && _dumped[i + 1] != '\n'
+					&& (_dumped[i + 1] == ']' || _dumped[i + 1] == '}'))
+					reformatted << '\n' << std::string(indentLevel * _format.indent, ' ');
+			}
+		}
+		reformatted << c;
+		if (!emptyThing && !inQuotes && (c == '[' || c == '{') && indentLevel > 0 && i + 1 < _dumped.size()
+			&& _dumped[i + 1] != '\n')
+			reformatted << '\n' << std::string(indentLevel * _format.indent, ' ');
+	}
+	return trimRightAllLines(reformatted.str());
 }
 
-/// Takes a JSON value (@ _json) and removes all its members with value 'null' recursively.
-void removeNullMembersHelper(Json::Value& _json)
+std::string escapeNewlinesAndTabsWithinStringLiterals(std::string const& _json)
 {
-	if (_json.type() == Json::ValueType::arrayValue)
-		for (auto& child: _json)
-			removeNullMembersHelper(child);
-	else if (_json.type() == Json::ValueType::objectValue)
-		for (auto const& key: _json.getMemberNames())
+	std::stringstream fixed;
+	bool inQuotes = false;
+	for (size_t i = 0; i < _json.size(); ++i)
+	{
+		char c = _json[i];
+
+		// Originally we had just this here:
+		// if (c == '"' && (i == 0 || _json[i - 1] != '\\'))
+		//    inQuotes = !inQuotes;
+		// However, this is not working if the escape character itself was escaped. e.g. "\n\r'\"\\".
+
+		if (c == '"')
 		{
-			Json::Value& value = _json[key];
-			if (value.isNull())
-				_json.removeMember(key);
+			size_t backslashCount = 0;
+			size_t j = i;
+			while (j > 0 && _json[j - 1] == '\\')
+			{
+				backslashCount++;
+				j--;
+			}
+			if (backslashCount % 2 == 0)
+			{
+				inQuotes = !inQuotes;
+				fixed << c;
+				continue;
+			}
+		}
+
+		if (inQuotes)
+		{
+			if (c == '\n')
```

Comment:
should we also support the other possible escape sequences?

![json-string-escape-diagram](https://github.com/ethereum/solidity/assets/5008794/da446d57-3045-464b-a43d-b81c1d0e66bc)